File: leetcode/0023-merge-k-sorted-lists.php
<?php

/**
 * 23. Merge k Sorted Lists
 * https://leetcode.com/problems/merge-k-sorted-lists/
 *
 * TODO: Runtime: 256 ms, faster than 36.59% of PHP online submissions for Merge k Sorted Lists.
 */

/**
 * Definition for a singly-linked list.
 * class ListNode {
 *     public $val = 0;
 *     public $next = null;
 *     function __construct($val = 0, $next = null) {
 *         $this->val = $val;
 *         $this->next = $next;
 *     }
 * }
 */
class Solution
{

    private $values = [];

    function addValues($node)
    {
        if ($node->val !== null) $this->values[] = $node->val;
        if ($node->next !== null) $this->addValues($node->next);
    }

    function getNodeList()
    {
        if (empty($this->values)) return null;
        return new ListNode(array_shift($this->values), $this->getNodeList());
    }

    /**
     * @param ListNode[] $lists
     * @return ListNode
     */
    function mergeKLists($lists)
    {
        foreach ($lists as $node)
            $this->addValues($node);
        sort($this->values);
        return $this->getNodeList();
    }
}

File: swift/0023-merge-k-sorted-lists.swift
/**
 * 23. Merge k Sorted Lists
 * https://leetcode.com/problems/merge-k-sorted-lists/
 **/

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public var val: Int
 *     public var next: ListNode?
 *     public init() { self.val = 0; self.next = nil; }
 *     public init(_ val: Int) { self.val = val; self.next = nil; }
 *     public init(_ val: Int, _ next: ListNode?) { self.val = val; self.next = next; }
 * }
 */
class Solution {
    func mergeKLists(_ lists: [ListNode?]) -> ListNode? {
        let head: ListNode? = ListNode()
        var cur: ListNode? = head, lists = lists

        while lists.count > 0 {
            while lists.count > 0 && lists[lists.count-1] == nil { lists.popLast() }
            if lists.isEmpty { break }

            var nextIndex: Int?, i = 0
            while i < lists.count {
                if lists[i] != nil {
                    if nextIndex == nil || lists[i]!.val < lists[nextIndex!]!.val {
                        nextIndex = i
                    }
                }
                i += 1
            }

            cur!.next = lists[nextIndex!]
            cur = cur?.next
            lists[nextIndex!] = lists[nextIndex!]!.next
        }

        return head!.next
    }
}